compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -464,11 +467,150 @@ fn check_opaque_meets_bounds<'tcx>(
         }
     }
-    // Clean up after ourselves
-    let _ = infcx.take_opaque_types();
+    // Check that any hidden types found during wf checking match the hidden types that `type_of` sees.
+    for (key, mut ty) in infcx.take_opaque_types() {
+        ty.hidden_type.ty = infcx.resolve_vars_if_possible(ty.hidden_type.ty);
+        sanity_check_found_hidden_type(tcx, key, ty.hidden_type, defining_use_anchor, origin)?;
+    }
     Ok(())
 }
 
+fn sanity_check_found_hidden_type<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: ty::OpaqueTypeKey<'tcx>,
+    mut ty: ty::OpaqueHiddenType<'tcx>,
+    defining_use_anchor: LocalDefId,
+    origin: &hir::OpaqueTyOrigin,
+) -> Result<(), ErrorGuaranteed> {
+    if ty.ty.is_ty_var() {
+        // Nothing was actually constrained.
+        return Ok(());
+    }
+    if let ty::Alias(ty::Opaque, alias) = ty.ty.kind() {
+        if alias.def_id == key.def_id.to_def_id() && alias.args == key.args {
+            // Nothing was actually constrained, this is an opaque usage that was
+            // only discovered to be opaque after inference vars resolved.
+            return Ok(());
+        }
+    }
+    // Closures frequently end up containing erased lifetimes in their final representation.
+    // These correspond to lifetime variables that never got resolved, so we patch this up here.
+    ty.ty = ty.ty.fold_with(&mut BottomUpFolder {
+        tcx,
+        ty_op: |t| t,
+        ct_op: |c| c,
+        lt_op: |l| match l.kind() {
+            RegionKind::ReVar(_) => tcx.lifetimes.re_erased,
+            _ => l,
+        },
+    });
+    // Get the hidden type.
+    let mut hidden_ty = tcx.type_of(key.def_id).instantiate(tcx, key.args);
+    // In case it is in a nested opaque type, find that opaque type's
+    // usage in the function signature and use the generic arguments from the usage site.
+    // We need to do because RPITs ignore the lifetimes of the function,
+    // as they have their own copies of all the lifetimes they capture.
+    // So the only way to get the lifetimes represented in terms of the function,
+    // is to look how they are used in the function signature (or do some other fancy
+    // recording of this mapping at ast -> hir lowering time).
+    if let hir::OpaqueTyOrigin::FnReturn(..) | hir::OpaqueTyOrigin::AsyncFn(..) = origin {
+        if hidden_ty != ty.ty {
+            hidden_ty = find_and_apply_rpit_substs(
+                tcx,
+                hidden_ty,
+                defining_use_anchor.to_def_id(),
+                key.def_id.to_def_id(),
+            )?;
+        }
+    }
+
+    // If the hidden types differ, emit a type mismatch diagnostic.
+    if hidden_ty == ty.ty {
+        Ok(())
+    } else {
+        let span = tcx.def_span(key.def_id);
+        let other = ty::OpaqueHiddenType { ty: hidden_ty, span };
+        Err(ty.report_mismatch(&other, key.def_id, tcx).emit())
+    }
+}
+
+fn find_and_apply_rpit_substs<'tcx>(
```

Comment:
can you add a doc comment which a minimal example of what this remapping does? something like

```rust
trait Id {
    type Assoc;
}
impl<'a> Id for &'a () {
    type Assoc = &'a ();
} 
fn func<'a>(x: &'a ()) -> impl Id<Assoc = impl Sized + 'a> { x }
// desugared to

type Outer<'a, 'aDup> = impl Id<Assoc = Inner<'a, 'aDup>>;
type Inner<'a, 'aDup> = impl Sized + 'aDup;
fn func<'a>(x: &'a () -> Outer<'a, 'a> { x }
```
- the type of the nested rpit is `&'a ()`
- when checking that `Outer` is wf, we reconstrain `Inner` to `&'aDup ()` which is different
- have to walk the signature of `func` to instantiate `'aDup` with `'a`

Reply:
done